ot/mpz-ot-core/src/ferret/cuckoo.rs

```diff
@@ -0,0 +1,195 @@
+    hashes: Arc<[AesEncryptor; CUCKOO_HASH_NUM]>,
+}
+
+impl CuckooHash {
+    /// Creates a new instance.
+    #[inline]
+    pub fn new(hashes: Arc<[AesEncryptor; CUCKOO_HASH_NUM]>) -> Self {
+        Self { hashes }
+    }
+
+    /// Insert elements into a Cuckoo hash table.
+    ///
+    /// * Argument
+    ///
+    /// * `alphas` - A u32 vector being inserted.
+    #[inline]
+    pub fn insert(&self, alphas: &[u32]) -> Result<Vec<Option<Item>>, CuckooHashError> {
+        // Always sets m = 1.5 * t. t is the length of `alphas`.
+        let m = compute_table_length(alphas.len() as u32);
+
+        // Allocates table.
+        let mut table = vec![None; m];
+        // Inserts each alpha.
+        for &value in alphas {
+            self.hash(&mut table, value)?
+        }
+        Ok(table)
+    }
+
+    // Hash an element to a position with the current hash function.
+    #[inline]
+    fn hash(&self, table: &mut [Option<Item>], value: u32) -> Result<(), CuckooHashError> {
+        // The item consists of the value and hash index, starting from 0.
+        let mut item = Item {
+            value,
+            hash_index: 0,
+        };
+
+        for _ in 0..CUCKOO_TRIAL_NUM {
+            // Computes the position of the value.
+            let pos = hash_to_index(&self.hashes[item.hash_index], table.len(), item.value);
+
+            // Inserts the value to position `pos`.
+            let opt_item = table[pos].replace(item);
+
+            // If position `pos` is not empty before the above insertion, iteratively inserts the obtained value.
+            if let Some(x) = opt_item {
+                item = x;
+                item.hash_index = (item.hash_index + 1) % CUCKOO_HASH_NUM;
+            } else {
+                // If no value assigned to position `pos`, end the process.
+                return Ok(());
+            }
+        }
+        Err(CuckooHashError)
+    }
+}
+
+/// Implementation of Bucket. See step 3 in Figure 7
+pub struct Bucket {
+    // The hash functions.
+    hashes: Arc<[AesEncryptor; CUCKOO_HASH_NUM]>,
+    // The number of buckets.
+    m: usize,
+}
+
+impl Bucket {
+    /// Creates a new instance.
+    #[inline]
+    pub fn new(hashes: Arc<[AesEncryptor; CUCKOO_HASH_NUM]>, m: usize) -> Self {
+        Self { hashes, m }
+    }
+
+    /// Inserts the input vector [0..n-1] into buckets.
+    ///
+    /// # Argument
+    ///
+    /// * `n` - The length of the vector [0..n-1].
+    #[inline]
+    pub fn insert(&self, n: u32) -> Vec<Vec<Item>> {
+        let mut buckets = vec![Vec::default(); self.m];
+        // NOTE: the sorted step in Step 3.c can be removed.
+        for i in 0..n {
+            for (index, hash) in self.hashes.iter().enumerate() {
+                let pos = hash_to_index(hash, self.m, i);
+                buckets[pos].push(Item {
+                    value: i,
+                    hash_index: index,
+                });
+            }
+        }
+        buckets
+    }
+}
```

Comment:
Similar reasons as I commented in the `CuckooHash` struct.

The functions between lines 173 and 185 should not be here, they are specific to the MPCOT protocol. We actually should make `CuckooHash` and `Bucket` independent of the MPCOT protocol. They can be used for other protocols such as PSI.

Reply:
I think we have to expose `Vec<Option<Item>>` and `Vec<Vec<Item>>` if cuckoo hash and bucket are taken as independent building blocks because they are not just for MPCOT, could be used for other cases.

So we can not encapsulate all these inside.